RecoMET/METAlgorithms/src/CSCHaloAlgo.cc

```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+  bool rhmatchingfound =false;
+  for(size_t i =0; i<haloclustercands.size(); i++){
+   
+    if(!ApplyMatchingCuts("EB",ishlt, haloclustercands[i].GetSeedEt(), iZ, haloclustercands[i].GetSeedZ(),iR, haloclustercands[i].GetSeedR(), iT,haloclustercands[i].GetSeedTime(), iPhi, haloclustercands[i].GetSeedPhi()))continue;
+    rhmatchingfound=true;
+    edm::RefVector<EcalRecHitCollection> bhrhcandidates = haloclustercands[i].GetBeamHaloRecHitsCandidates();
+    AddtoBeamHaloEBEERechits(bhrhcandidates, thehalodata,true);
+  }
+  return rhmatchingfound;
+}
+
+
+bool CSCHaloAlgo::SegmentMatchingEE(reco::CSCHaloData & thehalodata, std::vector<HaloClusterCandidateEE>haloclustercands, float iZ, float iR, float iT, float iPhi, bool ishlt){
+  bool rhmatchingfound =false;
+  for(size_t i =0; i<haloclustercands.size(); i++){
+   
+    if(!ApplyMatchingCuts("EE",ishlt, haloclustercands[i].GetSeedEt(), iZ, haloclustercands[i].GetSeedZ(),iR, haloclustercands[i].GetSeedR(), iT,haloclustercands[i].GetSeedTime(), iPhi, haloclustercands[i].GetSeedPhi()))continue;
+    rhmatchingfound=true;
+    edm::RefVector<EcalRecHitCollection> bhrhcandidates = haloclustercands[i].GetBeamHaloRecHitsCandidates();
+    AddtoBeamHaloEBEERechits(bhrhcandidates, thehalodata,false);
+  }
+  return rhmatchingfound;
+}
+
+bool CSCHaloAlgo::SegmentMatchingHB(reco::CSCHaloData & thehalodata, std::vector<HaloClusterCandidateHB>haloclustercands, float iZ, float iR, float iT, float iPhi, bool ishlt){
+  bool rhmatchingfound =false;
+  for(size_t i =0; i<haloclustercands.size(); i++){
+   
+    if(!ApplyMatchingCuts("HB",ishlt, haloclustercands[i].GetSeedEt(), iZ, haloclustercands[i].GetSeedZ(),iR, haloclustercands[i].GetSeedR(), iT,haloclustercands[i].GetSeedTime(), iPhi, haloclustercands[i].GetSeedPhi()))continue;
+    rhmatchingfound=true;
+    edm::RefVector<HBHERecHitCollection> bhrhcandidates = haloclustercands[i].GetBeamHaloRecHitsCandidates();
+    AddtoBeamHaloHBHERechits(bhrhcandidates, thehalodata);
+  }
+  return rhmatchingfound;
+}
+
+bool CSCHaloAlgo::SegmentMatchingHE(reco::CSCHaloData & thehalodata, std::vector<HaloClusterCandidateHE>haloclustercands, float iZ, float iR, float iT, float iPhi, bool ishlt){
+  bool rhmatchingfound =false;
+  for(size_t i =0; i<haloclustercands.size(); i++){
+   
+    if(!ApplyMatchingCuts("HE",ishlt, haloclustercands[i].GetSeedEt(),  iZ, haloclustercands[i].GetSeedZ(),iR, haloclustercands[i].GetSeedR(), iT,haloclustercands[i].GetSeedTime(), iPhi, haloclustercands[i].GetSeedPhi()))continue;
+    rhmatchingfound=true;
+    edm::RefVector<HBHERecHitCollection> bhrhcandidates = haloclustercands[i].GetBeamHaloRecHitsCandidates();
+    AddtoBeamHaloHBHERechits(bhrhcandidates, thehalodata);
+  }
+  return rhmatchingfound;
+}
+
+
+bool CSCHaloAlgo::ApplyMatchingCuts(TString subdet, bool ishlt, double rhet, double segZ, double rhZ, double segR, double rhR, double segT, double rhT, double segPhi, double rhPhi){
```

Comment:
I see that this method is used only with the same inputs from haloclustercands.
It may be more compact to pass a haloclustercand by const reference, instead of extracting the same 5 data members in multiple places.